async_producer.go

```diff
@@ -164,19 +154,19 @@ func (pe ProducerErrors) Error() string {
 	return fmt.Sprintf("kafka: Failed to deliver %d messages.", len(pe))
 }
 
-func (p *producer) Errors() <-chan *ProducerError {
+func (p *asyncProducer) Errors() <-chan *ProducerError {
 	return p.errors
 }
 
-func (p *producer) Successes() <-chan *ProducerMessage {
+func (p *asyncProducer) Successes() <-chan *ProducerMessage {
 	return p.successes
 }
 
-func (p *producer) Input() chan<- *ProducerMessage {
-	return p.input
+func (p *asyncProducer) EnqueueMessage(msg *ProducerMessage) {
```

Comment:
In fact we basically have to make them callbacks or else we end up with deadlocks when writing the intuitive pattern (which is to call `EnqueueMessage` in the `default` case of a select)